lib/posix/pthread.c

```diff
@@ -268,15 +265,14 @@ int pthread_cancel(pthread_t pthread)
 	k_spin_unlock(&thread->cancel_lock, cancel_key);
 
 	if (cancel_state == PTHREAD_CANCEL_ENABLE) {
-		pthread_mutex_lock(&thread->state_lock);
+		state_key = k_spin_lock(&thread->state_lock);
 		if (thread->state == PTHREAD_DETACHED) {
 			thread->state = PTHREAD_TERMINATED;
```

Comment:
yeah, this looks like a locking issue -- you're setting the state to TERMINATED while holding the spinlock, but the above allocation code isn't holding the lock when transitioning from TERMINATED to a running state. I think you need to either do all state manipulations using atomics or all state manipulations under the spin lock. And, I think the latter option will be easier to reason about.

Reply:
Hmm.. yea, I could see switching to only spinlocks.